### PR TITLE
Add support for `filter_roles` option when listing flows and runs

### PR DESCRIPTION
**Pull request by m1yag1**
https://app.shortcut.com/globus/story/40419
### Added
 - `FlowsClient.list_flows` and `FlowsClient.list_runs` now support the
   `filter_roles` parameter to filter results by one or more roles.

----
📚 Documentation preview 📚: https://globus-sdk-python--1174.org.readthedocs.build/en/1174/

### CODE REVIEW

File: src/globus_sdk/services/flows/client.py [hunk 1]
```diff
@@ -260,6 +265,16 @@ def list_flows(
         the ``flow_starter``, ``flow_administrator`` or ``flow_owner`` roles will be
         returned.
 
+        The role values ``run_manager`` and ``run_monitor`` are also supported but do
+        not follow an order of precedence.
+
+        **Filter Roles Values**
+
+        The valid values for ``filter_roles`` are the same as those for ``filter_role``.
```
Comment (m1yag1) [inline src/globus_sdk/services/flows/client.py:273]: I wasn't sure how to describe the differences between the `filter_role` and `filter_roles`. I tried to avoid using the term' legacy" and wanted to treat them as separate things. The addition of the `run_managers` and `run_monitors` not being part of the precedence order also made this a bit awkward. I'm open to any suggestions on how to make this documentation better.

**Review by sirosen — CHANGES_REQUESTED** [review]
We primarily need one minor fix: comma-joining the lists.

I've also left some notes on doc improvements and handling of empty strings and empty lists, which I think are worth our while, but the comma-joining is the only blocker I see here.

File: src/globus_sdk/services/flows/client.py [hunk 2]
```diff
@@ -327,8 +342,15 @@ def list_flows(
             query_params = {}
         if filter_role is not None:
             query_params["filter_role"] = filter_role
+        if filter_roles is not None:
+            query_params["filter_roles"] = filter_roles
         if filter_fulltext is not None:
             query_params["filter_fulltext"] = filter_fulltext
+
+        if filter_role and filter_roles:
```
[the remark this answers is not in the document]
Comment (m1yag1) [reply src/globus_sdk/services/flows/client.py]: I updated this and added those values to the mutually exclusive test case.

File: tests/functional/services/flows/test_list_flows.py [hunk 3]
```diff
@@ -109,3 +110,25 @@ def test_list_flows_orderby_multi(flows_client, orderby_style, orderby_value):
...
+
+    res = flows_client.list_flows(filter_roles=filter_roles)
+    assert res.http_status == 200
+
+    req = get_last_request()
+    assert req.body is None
+    parsed_qs = urllib.parse.parse_qs(urllib.parse.urlparse(req.url).query)
+
+    if filter_roles:
+        assert parsed_qs["filter_roles"] == filter_roles
```
[the remark this answers is not in the document]
Comment (m1yag1) [reply tests/functional/services/flows/test_list_flows.py]: I added updated changes to this test. I hope I understood correctly and attempted to add parametrized values that tested the comma-joining logic.

File: src/globus_sdk/services/flows/client.py [hunk 4]
```diff
@@ -327,8 +342,15 @@ def list_flows(
             query_params = {}
         if filter_role is not None:
             query_params["filter_role"] = filter_role
+        if filter_roles is not None:
+            query_params["filter_roles"] = filter_roles
```
[the remark this answers is not in the document]
Comment (m1yag1) [reply src/globus_sdk/services/flows/client.py]: @sirosen I updated these usage sites with `utils.commajoin()`.

File: tests/functional/services/flows/test_list_runs.py [hunk 5]
```diff
@@ -68,3 +69,37 @@ def test_list_runs_filter_flow_id(flows_client, pass_as_uuids):
...
+
+@pytest.mark.parametrize(
+    "filter_roles, expected_filter_roles",
+    [
+        # empty list
+        ([], []),
+        # list with empty string
+        ([""], [""]),
+        # None
+        (None, None),
```
Comment (sirosen) [inline tests/functional/services/flows/test_list_runs.py:82]: The way this was written actually taught me a bit of corner-case behavior in `requests` that I didn't know about. I didn't realize that passing an empty string resulted in a parameter being _omitted_. (Remember that we're going through `utils.commajoin`)
However, I think the way the test is written gives a different impression about what's being asserted.

I tested this rewrite locally and it does the same thing (and passes):
```suggestion
        # empty list, list with empty string, and None do not send the param
        ([], None),
        ([""], None),
        (None, None),
```

I think this is much better at communicating what the test asserts -- the variable right-hand-side values in these parameters makes it look like we're testing three different outcomes. Making them all `None` clarifies that we're testing for the same outcome.

File: tests/functional/services/flows/test_list_runs.py [hunk 6]
```diff
@@ -68,3 +69,37 @@ def test_list_runs_filter_flow_id(flows_client, pass_as_uuids):
...
+def test_list_runs_filter_roles(flows_client, filter_roles, expected_filter_roles):
+    load_response(flows_client.list_runs).metadata
+    res = flows_client.list_runs(filter_roles=filter_roles)
+    assert res.http_status == 200
+
+    req = get_last_request()
+    assert req.body is None
+    parsed_qs = urllib.parse.parse_qs(urllib.parse.urlparse(req.url).query)
+
+    if expected_filter_roles and any(role for role in expected_filter_roles):
```
Comment (sirosen) [inline tests/functional/services/flows/test_list_runs.py:102]: The above also lets us simplify this check:
```suggestion
    if expected_filter_roles:
```

The `any()` check was what initially drew my attention -- it seemed odd but passes when run.

File: tests/functional/services/flows/test_list_flows.py [hunk 7]
```diff
@@ -109,3 +110,70 @@ def test_list_flows_orderby_multi(flows_client, orderby_style, orderby_value):
...
+
+    req = get_last_request()
+    assert req.body is None
+    parsed_qs = urllib.parse.parse_qs(urllib.parse.urlparse(req.url).query)
+
+    # Only check filter_roles in parsed_qs if we expect a non-empty value
+    if expected_filter_roles and any(role for role in expected_filter_roles):
+        assert parsed_qs["filter_roles"] == expected_filter_roles
+    else:
+        assert "filter_roles" not in parsed_qs
```
Comment (sirosen) [inline tests/functional/services/flows/test_list_flows.py:179]: I'd like to see this test updated in a way that accords with my recommendation on the `list_runs` test, for the same reasons -- declaring multiple "expected results" which are all equivalent gives a reader the wrong impression when going through the parameters.

File: src/globus_sdk/services/flows/client.py [hunk 8]
```diff
@@ -240,25 +241,39 @@ def list_flows(
         List deployed flows
 
         :param filter_role: A role name specifying the minimum permissions required for
-            a flow to be included in the response.
+            a flow to be included in the response. Mutually exclusive with
+            **filter_roles**.
```
Comment (sirosen) [inline src/globus_sdk/services/flows/client.py:245]: As a very minor tweak, I think we can call this deprecated here, as a hint to users that they should use `filter_roles` instead:
```suggestion
            **filter_roles**. (deprecated)
```

File: src/globus_sdk/services/flows/client.py [hunk 9]
```diff
@@ -327,8 +342,15 @@ def list_flows(
             query_params = {}
         if filter_role is not None:
             query_params["filter_role"] = filter_role
```
Comment (sirosen) [inline src/globus_sdk/services/flows/client.py:344]: We discussed adding a deprecation warning here (we have tooling for this, use the `warn_deprecated` helper), which I think we're on solid ground to include. It's not clear when this is being removed from the API, but adding the warning gives us better grounds to remove the parameter in v4 and have a simpler SDK interface.

**Review by sirosen — CHANGES_REQUESTED** [review]

File: tests/functional/services/flows/test_list_flows.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sirosen) [reply tests/functional/services/flows/test_list_flows.py]: Yeah, I'm very happy with the new tests! I have an adjustment I'd like us to make -- noted in a review -- but the initial parametrization put us on good footing to discuss that.

File: src/globus_sdk/services/flows/client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sirosen) [reply src/globus_sdk/services/flows/client.py]: Yep, I think the `src/` tree stuff is now all good on this account!

File: src/globus_sdk/services/flows/client.py (hunk 1 above)
Comment (sirosen) [reply src/globus_sdk/services/flows/client.py]: Thanks! I tried to give what I thought was a pretty good suggestion, so I'm happy with you taking it. :grin:

File: src/globus_sdk/services/flows/client.py (hunk 8 above)
Comment (m1yag1) [reply src/globus_sdk/services/flows/client.py:252]: I applied this and it was in the wrong place (filter_roles) but i fixed it 🙃.
